src/did-manager/cheqd-did-provider.ts

```diff
@@ -104,16 +109,336 @@ export type TImportableEd25519Key = Required<Pick<IKey, 'publicKeyHex' | 'privat
+			}
+			// Check if controller DID document contains verification methods
+			if (!controllerDidDocument.verificationMethod) {
+				throw new CheqdProviderError(
+					'[did-provider-cheqd]: compileSignInfos: Controller DID document does not contain verification methods',
+					CheqdProviderErrorCodes.EmptyVerificationMethod);
+			}
+			// Iterate over verificationMethods and by comparing publicKeys get the verificationMethod Id
+			for (const vm of controllerDidDocument.verificationMethod) {
+				// Try to match verificationMethod with one of publicKeyHexs
+				const verificationMethodId = ( function (){
+					for (const publicKeyHex of publicKeyHexs) {
+						// Transform to string
+						const publicKey = fromString(publicKeyHex, 'hex');
+						switch (vm?.type) {
+							case VerificationMethods.Ed255192020: {
+								const publicKeyMultibase = toMultibaseRaw(publicKey);
+								if (vm.publicKeyMultibase === publicKeyMultibase) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+							case VerificationMethods.Ed255192018: {
+								const publicKeyBase58 = bases['base58btc'].encode(publicKey).slice(1);
+								if (vm.publicKeyBase58 === publicKeyBase58) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+							case VerificationMethods.JWK: {
+								const publicKeyJwk: JsonWebKey = {
+									crv: 'Ed25519',
+									kty: 'OKP',
+									x: toString(publicKey, 'base64url'),
+								};
+								if (JSON.stringify(vm.publicKeyJwk) === JSON.stringify(publicKeyJwk)) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+						}
+					}
+				})()
+
+				if (verificationMethodId) {
+					verificationMethodIds.push(verificationMethodId);
+				}
+				// Setup key structure for display
+				const kid = extractPublicKeyHex(vm);
+				const key = await this.context.agent.keyManagerGet({ kid });
+				this.controllerKeys.push({...key, controller: vm.controller} satisfies ICheqdKey);
+			}
+		}
+
+		// Iterate over pair of verificationMethodIds and publicKeys and create SignInfo
+		for (const { verificationMethodId, publicKeyHex } of verificationMethodIds) {
+			signInfos.push({
+				verificationMethodId,
+				signature: base64ToBytes(
+					await this.context.agent.keyManagerSign({
+						keyRef: publicKeyHex,
+						data: toString(payload, 'hex'),
+						encoding: 'hex',
+					})
+				),
+			} satisfies SignInfo)
+		}
+		this.setSignInfos(signInfos)
+	}
+
+	async updateIdentifierCompileSignInfos(
+		didDocument: DIDDocument, 
+		options: {
+			versionId?: string
+			publicKeyHexs?: string[];
+		}
+		): Promise<void> {
+
+		// Steps to solve the issue:
+		// 1. Collect list of controllers. The user can remove, append and reqrite the controller.
```

Comment:
This looks a bit ambiguous.

Is the instruction here including all allowed operations as in update, add, remove controllers?

Reply:
Yup. Updating DID Document means that it can be any kind of operations like adding, replacing or removing at all.